File: client-kotlin/app/src/main/java/ua/meugen/android/rxjavatests_client/presenter/MainPresenter.java
package ua.meugen.android.rxjavatests_client.presenter;

import android.util.Log;

import com.hannesdorfmann.mosby3.mvp.MvpPresenter;

import javax.inject.Inject;

import rx.Observable;
import rx.Subscription;
import rx.subscriptions.CompositeSubscription;
import ua.meugen.android.rxjavatests_client.model.responses.DataResponse;
import ua.meugen.android.rxjavatests_client.presenter.rest.ModelApi;
import ua.meugen.android.rxjavatests_client.view.MainView;

/**
 * @author meugen
 */

public class MainPresenter implements MvpPresenter<MainView> {

    private static final String TAG = MainPresenter.class.getSimpleName();

    private static final long DELAY = 5000L;
    private static final int COUNT = 100;

    private final ModelApi modelApi;

    private MainView view;
    private CompositeSubscription compositeSubscription;

    @Inject
    public MainPresenter(final ModelApi modelApi) {
        this.modelApi = modelApi;
    }

    @Override
    public void attachView(final MainView view) {
        this.view = view;
        compositeSubscription = new CompositeSubscription();
    }

    @Override
    public void detachView(final boolean retainInstance) {
        this.view = null;
        if (compositeSubscription != null) {
            compositeSubscription.unsubscribe();
            compositeSubscription = null;
        }
    }

    public void setup(final boolean reload) {
        view.showProgress();
        Observable<DataResponse> observable = modelApi.getDataObservable();
        if (reload || observable == null) {
            observable = modelApi.dataWithDelay(DELAY, COUNT);
        }
        Subscription subscription = observable
                .subscribe(this::onDataSucess, this::onDataError);
        compositeSubscription.add(subscription);
    }

    public void onDataError(final Throwable e) {
        Log.e(TAG, e.getMessage(), e);
    }

    public void onDataSucess(final DataResponse response) {
        Log.i(TAG, "onDataSucess(" + response + ")");
        view.showData(response.getData());
    }
}

File: client-kotlin/app/src/main/java/ua/meugen/android/rxjavatests_client/presenter/MainPresenter.kt
package ua.meugen.android.rxjavatests_client.presenter

import android.util.Log

import com.hannesdorfmann.mosby3.mvp.MvpPresenter

import javax.inject.Inject

import rx.Observable
import rx.subscriptions.CompositeSubscription
import ua.meugen.android.rxjavatests_client.model.DataResponse
import ua.meugen.android.rxjavatests_client.presenter.rest.ModelApi
import ua.meugen.android.rxjavatests_client.view.MainView

/**
 * @author meugen
 */

class MainPresenter @Inject constructor(private val modelApi: ModelApi) : MvpPresenter<MainView> {

    private var view: MainView? = null
    private var compositeSubscription: CompositeSubscription? = null

    override fun attachView(view: MainView) {
        this.view = view
        compositeSubscription = CompositeSubscription()
    }

    override fun detachView(retainInstance: Boolean) {
        this.view = null
        compositeSubscription?.unsubscribe()
        compositeSubscription = null;
    }

    fun setup(reload: Boolean) {
        view!!.showProgress()
        var observable: Observable<DataResponse>? = modelApi.dataObservable
        if (reload || observable == null) {
            observable = modelApi.dataWithDelay(DELAY, COUNT)
        }
        val subscription = observable
                .subscribe({ this.onDataSucess(it) }, { this.onDataError(it) })
        compositeSubscription!!.add(subscription)
    }

    fun onDataError(e: Throwable) {
        Log.e(TAG, e.message, e)
    }

    fun onDataSucess(response: DataResponse) {
        Log.i(TAG, "onDataSucess($response)")
        view!!.showData(response.data!!)
    }

    companion object {

        private val TAG = MainPresenter::class.java.simpleName

        private val DELAY = 5000L
        private val COUNT = 100
    }
}
